libraries/rush-lib/src/logic/operations/BuildPlanPlugin.ts

```diff
@@ -0,0 +1,298 @@
+  clusters: Set<Operation>[];
+  buildCacheByOperation: Map<Operation, IBuildPlanOperationCacheContext>;
+  clusterByOperation: Map<Operation, Set<Operation>>;
+}
+
+export class BuildPlanPlugin implements IPhasedCommandPlugin {
+  private readonly _terminal: ITerminal;
+
+  public constructor(terminal: ITerminal) {
+    this._terminal = terminal;
+  }
+
+  public apply(hooks: PhasedCommandHooks): void {
+    const terminal: ITerminal = this._terminal;
+    hooks.beforeExecuteOperations.tapPromise(PLUGIN_NAME, createBuildPlan);
+
+    async function createBuildPlan(
+      recordByOperation: Map<Operation, IOperationExecutionResult>,
+      context: IExecuteOperationsContext
+    ): Promise<void> {
+      const { projectConfigurations, projectChangeAnalyzer } = context;
+      const disjointSet: DisjointSet<Operation> = new DisjointSet<Operation>();
+      const operations: IterableIterator<Operation> = recordByOperation.keys();
+      for (const operation of operations) {
+        disjointSet.add(operation);
+      }
+      const buildCacheByOperation: Map<Operation, IBuildPlanOperationCacheContext> = new Map<
+        Operation,
+        IBuildPlanOperationCacheContext
+      >();
+      for (const operation of operations) {
+        const { associatedProject, associatedPhase } = operation;
+        if (associatedProject && associatedPhase) {
+          const projectConfiguration: RushProjectConfiguration | undefined =
+            projectConfigurations.get(associatedProject);
+          const fileHashes: Map<string, string> | undefined =
+            await projectChangeAnalyzer._tryGetProjectDependenciesAsync(associatedProject, terminal);
+          const cacheDisabledReason: string | undefined = projectConfiguration
+            ? projectConfiguration.getCacheDisabledReason(fileHashes!.keys(), associatedPhase.name, operation)
+            : `Project does not have a ${RushConstants.rushProjectConfigFilename} configuration file, ` +
+              'or one provided by a rig, so it does not support caching.';
+          buildCacheByOperation.set(operation, { cacheDisabledReason });
+        }
+      }
+      clusterOperations(disjointSet, buildCacheByOperation);
+      const buildPlan: ICobuildPlan = createCobuildPlan(disjointSet, terminal, buildCacheByOperation);
+      logCobuildBuildPlan(buildPlan, terminal);
+    }
+  }
+}
+
+function generateCobuildPlanSummary(operations: Operation[], terminal: ITerminal): ICobuildPlan['summary'] {
+  const noOpOperations: Set<Operation> = new Set(operations.filter((e) => e.runner?.isNoOp));
+  const leafQueue: Operation[] = [];
+
+  for (const operation of operations) {
+    if (noOpOperations.has(operation)) {
+      continue;
+    }
+    let numberOfConsumers: number = 0;
+    for (const consumer of operation.consumers) {
+      if (!noOpOperations.has(consumer)) {
+        numberOfConsumers++;
+      }
+    }
+    if (numberOfConsumers === 0) {
+      leafQueue.push(operation);
+    }
```

Comment:
Just because your only consumer is a noop does not mean that you are a leaf of the build tree. You are only a leaf if all of your consumers' consumers are noops all the way down.

A -> noop -> B is completely valid. The graph construction deliberately leaves in such nodes, since more typically that looks like:
```
A -\       /- E
B -- > D < -- F
C -/       \- G
```

Reply:
Good call, I updated this to a breadth-first traversal to build up the number of dependencies and then we assign those with no dependencies (no matter the depth) to the initial queue.